linkerd/opaque-transport/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+use bytes::{
+    buf::{Buf, BufMut},
+    Bytes, BytesMut,
+};
+use linkerd2_dns_name::Name;
+use linkerd2_error::Error;
+use linkerd2_io::{self as io, AsyncReadExt, AsyncWriteExt};
+use linkerd2_proxy_transport::Detect;
+use prost::Message;
+use std::str::FromStr;
+use tracing::trace;
+
+mod proto {
+    include!(concat!(env!("OUT_DIR"), "/opaque.proxy.l5d.io.rs"));
+}
+
+#[derive(Clone, Debug)]
+pub struct Header {
+    /// The target port.
+    pub port: u16,
+
+    /// The logical name of the target (service), if one is known.
+    pub name: Option<Name>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectHeader(());
+
+const PREFACE: &[u8] = b"proxy.l5d.io/opaque\r\n\r\n";
+const PREFACE_AND_SIZE_LEN: usize = PREFACE.len() + 4;
+
+#[async_trait::async_trait]
+impl Detect for DetectHeader {
+    type Protocol = Header;
+
+    #[inline]
+    async fn detect<I: io::AsyncRead + Send + Unpin + 'static>(
+        &self,
+        io: &mut I,
+        buf: &mut BytesMut,
+    ) -> Result<Option<Header>, Error> {
+        Header::read_prefaced(io, buf).await.map_err(Into::into)
+    }
+}
+
+impl Header {
+    pub async fn write(&self, io: &mut (impl io::AsyncWrite + Unpin)) -> Result<usize, Error> {
+        let mut buf = self.encode_prefaced_buf()?;
+        let mut sz = 0usize;
+
+        while !buf.is_empty() {
+            sz += io.write_buf(&mut buf).await?;
+            trace!(written = sz, remaining = buf.len(), "Wrote header");
+        }
+
+        Ok(sz)
+    }
+
+    #[inline]
+    pub fn encode_prefaced_buf(&self) -> Result<Bytes, Error> {
+        let mut buf = BytesMut::new();
+        self.encode_prefaced(&mut buf)?;
+        Ok(buf.freeze())
+    }
+
+    /// Encodes the connection header to a byte buffer.
+    pub fn encode_prefaced(&self, buf: &mut BytesMut) -> Result<(), Error> {
+        buf.reserve(PREFACE_AND_SIZE_LEN);
+        buf.put(PREFACE);
+
+        debug_assert!(buf.capacity() >= 4);
+        // Safety: These bytes must be initialized below once the message has
+        // been encoded.
+        unsafe {
+            buf.advance_mut(4);
+        }
```

Comment:
I'm not actually sure if the backfilling is necessary --- even if we don't use [`Message::encode_length_delimited`](https://docs.rs/prost/0.6.1/prost/trait.Message.html#method.encode_length_delimited) as I mentioned below, I think we could use [`Message::encoded_len`](https://docs.rs/prost/0.6.1/prost/trait.Message.html#tymethod.encoded_len) to get the length of the message _prior_ to encoding it?

Reply:
Again, this is uses a varint, which we dont' want to use.

Reply:
`encoded_len` returns a `usize`, so we could just assert it's less than `u32::max`, right? The varint encoding happens in `encode_length_delimited`; AFAICT we should still be able to use `encoded_len`?

Reply:
FWIW, this all comes directly from tonic's message encoding https://github.com/hyperium/tonic/blob/1b03ece2a81cb7e8b1922b3c3c1f496bd402d76c/tonic/src/codec/encode.rs#L54-L67. I'll look at using encoded_len though.

Reply:
oh, huh, i wonder why tonic does that. well, it could be a bit nicer if `encoded_len` works, but if there's some other reason we can't use it, nbd...